### PR TITLE
Allow option of exposing VM domain info via instance metadata

### PR DESCRIPTION
**Pull request by mlsorensen**
Signed-off-by: Marcus Sorensen <mls@apple.com>

### Description

This PR allows the cloud admin to set either a global or domain-specific value "metadata.allow.expose.domain", and when set this allows the VM to see the name and ID of the immediate domain that contains the VM in instance metadata. This can be useful or a variety of things such as bootstrapping VM configuration and access according to domain.

This PR also deletes the CloudZonesNetworkElement because it isn't referred to anywhere, and there was initially some confusion as to whether this code needed to be updated when extending metadata.  If it needs to be kept we can remove that delete from the PR.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

```
[root@test-vm ~]# cloud-init query ds.meta_data.cloud_domain
mydomain

[root@test-vm ~]# cloud-init query ds.meta_data.cloud_domain_id
e41e92fa-0498-4e76-b4b3-5948a44a9309
```

### How Has This Been Tested?
Tested in a local branch, released to an internal UAT environment for a few weeks. Also updated metadata Marvin test.

Tested both setting "metadata.allow.expose.domain" true and false, and going back and forth. Need to restart the VM for changes to take effect.

### CODE REVIEW

**Comment by mlsorensen** [comment]
Not sure why the code coverage is complaining on the appenders, seems unrelated to the PR.

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@DaanHoogland how can we run a specific component test with BO?

**Comment by DaanHoogland** [comment]
@rohityadavcloud we had disabled the feature in BO two or three years ago as it was never used. I'll look at reimplement it.

**Review by GutoVeronezi — APPROVED** [review]
LGTM

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java [hunk 1]
```diff
@@ -2551,6 +2553,10 @@ public List<String[]> generateVmData(String userData, String serviceOffering, lo
         final String zoneName = dcVo.getName();
 
         IPAddressVO publicIp = _ipAddressDao.findByAssociatedVmId(vmId);
+        VirtualMachine vm = _vmDao.findById(vmId);
+        if (vm == null) {
+            throw new CloudRuntimeException("Cannot generate VM instance data, no VM exists by given ID");
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:2558]:
```suggestion
            throw new CloudRuntimeException(String.format("Cannot generate VM instance data, no VM exists by ID: %d", vmId));
```

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java [hunk 2]
```diff
@@ -2619,6 +2625,14 @@ public List<String[]> generateVmData(String userData, String serviceOffering, lo
             vmData.add(new String[]{PASSWORD_DIR, PASSWORD_FILE, password});
         }
         vmData.add(new String[]{METATDATA_DIR, HYPERVISOR_HOST_NAME_FILE, hostname});
+
+        Domain domain = _domainDao.findById(vm.getDomainId());
+        if (domain != null && VirtualMachineManager.AllowExposeDomainInMetadata.valueIn(domain.getId())) {
+            s_logger.debug("Adding domain info to cloud metadata");
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:2631]: @GutoVeronezi could that be a possible security concern? having private user data in the logs.

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mlsorensen) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:2558]: Good idea. I think the cloudruntimeexception will be logged at some stage, committing Daan's suggestion.

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlsorensen) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:2631]: I think in context of the full job log this info would be apparent, someone troubleshooting would see what VM was being acted upon.

File: server/src/main/java/com/cloud/network/router/CommandSetupHelper.java [hunk 3]
```diff
@@ -210,11 +214,18 @@ public void createVmDataCommand(final VirtualRouter router, final UserVm vm, fin
...
-                    generateVmDataCommand(router, nic.getIPv4Address(), vm.getUserData(), serviceOffering, zoneName,
-                            staticNatIp == null || staticNatIp.getState() != IpAddress.State.Allocated ? null : staticNatIp.getAddress().addr(), vm.getHostName(), vm.getInstanceName(),
-                            vm.getId(), vm.getUuid(), publicKey, nic.getNetworkId(), destHostname));
+            VmDataCommand vmDataCommand = generateVmDataCommand(router, nic.getIPv4Address(), vm.getUserData(), serviceOffering, zoneName,
+                    staticNatIp == null || staticNatIp.getState() != IpAddress.State.Allocated ? null : staticNatIp.getAddress().addr(), vm.getHostName(), vm.getInstanceName(),
+                    vm.getId(), vm.getUuid(), publicKey, nic.getNetworkId(), destHostname);
+
+            Domain domain = domainDao.findById(vm.getDomainId());
+            if (domain != null && VirtualMachineManager.AllowExposeDomainInMetadata.valueIn(domain.getId())) {
+                s_logger.debug("Adding domain info to cloud metadata");
```
[the remark this answers is not in the document]
Comment (mlsorensen) [reply server/src/main/java/com/cloud/network/router/CommandSetupHelper.java:223]: I think in context of the individual job log this info would be apparent, someone troubleshooting (e.g. searching logs by job or logid) would know what VM was being acted upon.

**Review by DaanHoogland — APPROVED** [review]
CLGTM

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (GutoVeronezi) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:2631]: Regarding @DaanHoogland's comment, operators already have access to the VMs metadata (e.g.: while listing the VMs via API) and these data are printed several times in other logs; therefore, I do not see how could it be a security concern in this case. Am I missing something?

[the remark this answers is not in the document]
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:2631]: you are right @GutoVeronezi , my bad. I read userdata for metadata. I think we should prevent it, but am not insisting. As for @mlsorensen 's argument. I think it makes sense but would not hurt the other way around either, for trouble shooting.
+/-0

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]
LGTM, the one failing test is intermittent failure not caused by this PR

**Pull request merged**